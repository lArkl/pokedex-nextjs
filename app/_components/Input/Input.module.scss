.container {
  border-radius: 0.2rem;
  border: 1px solid black;
  padding: 0.55rem;

  &:hover {
    border-color: var(--color-gray-6);
  }
}
