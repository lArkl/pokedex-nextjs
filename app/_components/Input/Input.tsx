import { InputHTMLAttributes } from 'react'
import styles from './Input.module.scss'
import { Control, FieldValues, useController, FieldPath, UseControllerProps } from 'react-hook-form'
import classNames from 'classnames'

export type InputProps<TFieldValues extends FieldValues> = InputHTMLAttributes<HTMLInputElement> & {
  name: FieldPath<TFieldValues>
  control: Control<TFieldValues>
  rules?: UseControllerProps['rules']
}

export default function Input<TFieldValues extends FieldValues>({
  control,
  name,
  rules,
  ...props
}: InputProps<TFieldValues>) {
  const { field } = useController<TFieldValues>({
    control,
    name,
    rules,
  })
  return <input {...props} id={name} {...field} className={classNames([styles.container, props.className])} />
}
