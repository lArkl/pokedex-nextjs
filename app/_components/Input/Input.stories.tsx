import type { Meta, StoryObj } from '@storybook/react'

import Input from './Input'
import { useForm } from 'react-hook-form'
import { InputProps } from 'react-select'

const meta = {
  title: 'Atoms/Input',
  component: Input,
  parameters: {
    layout: 'centered',
  },
  tags: ['autodocs'],
  argTypes: {
    control: {
      control: false,
    },
  },
} satisfies Meta<typeof Input>

export default meta
type BaseStory = StoryObj<typeof meta>
type Story = Omit<BaseStory, 'args'> & { args: Omit<BaseStory['args'], 'control' | 'name'> }

const Component = (args: Partial<InputProps>) => {
  const { control } = useForm({ defaultValues: { username: args.defaultValue ?? '' } })
  return <Input {...args} name="username" control={control} />
}

export const Default: Story = {
  render: (args) => <Component {...args} />,
  args: {},
}

export const DateType: Story = {
  render: (args) => <Component {...args} />,
  args: {
    type: 'date',
    value: '2023-03-02',
  },
}
