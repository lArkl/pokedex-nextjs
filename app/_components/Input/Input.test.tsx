import { fireEvent, render, screen } from '@testing-library/react'
import Input, { InputProps } from './Input'
import { useForm } from 'react-hook-form'

const WrappedInput = ({ defaultValue, ...props }: Omit<InputProps<{ field: string }>, 'control' | 'name'> = {}) => {
  const { control } = useForm({ defaultValues: { field: defaultValue } })
  return (
    <form data-testid="form">
      <Input name="field" {...props} control={control} />
      <button>submit</button>
    </form>
  )
}

describe('Input', () => {
  it('renders component', async () => {
    const { asFragment } = render(<WrappedInput />)

    expect(asFragment()).toMatchSnapshot()
  })

  it('accepts a className', async () => {
    render(<WrappedInput className="class" />)

    const input = screen.getByRole('textbox')
    expect(input).toHaveClass('class')
  })

  it('should update value when typing', async () => {
    render(<WrappedInput />)

    const input = screen.getByRole('textbox')
    fireEvent.change(input, { target: { value: 'test' } })

    expect(input).toHaveValue('test')
  })

  it('should change value when typing', async () => {
    render(<WrappedInput />)
    expect(screen.getByTestId('form')).toHaveFormValues({ field: '' })

    const input = screen.getByRole('textbox')
    fireEvent.change(input, { target: { value: 'test' } })

    expect(screen.getByTestId('form')).toHaveFormValues({ field: 'test' })
  })
})
