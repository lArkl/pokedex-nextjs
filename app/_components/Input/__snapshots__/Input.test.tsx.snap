// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Input > renders component 1`] = `
<DocumentFragment>
  <form
    data-testid="form"
  >
    <input
      class="_container_859b20"
      id="field"
      name="field"
      value=""
    />
    <button>
      submit
    </button>
  </form>
</DocumentFragment>
`;
